json_rpc/clients/httpclient.nim

```diff
@@ -14,188 +16,45 @@ type
-      # everything until connection will be closed.
-      var blenfut = transp.read(maxBodySize)
-      let ores = await withTimeout(blenfut, HttpBodyTimeout)
-      if not ores:
-        # Timeout
-        debug "Timeout expired while receiving request body",
-              address = transp.remoteAddress()
-        return ""
-
-      buffer = blenfut.read()
-  except TransportIncompleteError:
-    # remote peer disconnected
-    debug "Remote peer disconnected", address = transp.remoteAddress()
-    return ""
-  except TransportOsError as exc:
-    debug "Problems with networking", address = transp.remoteAddress(),
-          error = exc.msg
-    return ""
-
-  return string.fromBytes(buffer)
 
 proc new(T: type RpcHttpClient, maxBodySize = MaxHttpRequestSize): T =
   T(
     maxBodySize: maxBodySize,
-    options: HttpClientOptions(httpMethod: MethodPost),
+    httpSession: HttpSessionRef.new(),
   )
 
 proc newRpcHttpClient*(maxBodySize = MaxHttpRequestSize): RpcHttpClient =
   RpcHttpClient.new(maxBodySize)
 
-proc httpMethod*(client: RpcHttpClient): HttpMethod =
-  client.options.httpMethod
-
-proc httpMethod*(client: RpcHttpClient, m: HttpMethod) =
-  client.options.httpMethod = m
-
 method call*(client: RpcHttpClient, name: string,
-             params: JsonNode): Future[Response] {.
-    async, gcsafe, raises: [Defect, CatchableError].} =
-  ## Remotely calls the specified RPC method.
-  let id = client.getNextId()
+             params: JsonNode): Future[Response]
+            {.async, gcsafe, raises: [Defect, CatchableError].} =
+  doAssert client.httpSession != nil
+  if client.httpAddress.isErr:
+    raise newException(RpcAddressUnresolvableError, client.httpAddress.error)
 
   let
-    transp = await connect(client.addresses[0])
+    id = client.getNextId()
     reqBody = $rpcCallNode(name, params, id)
-    res = await transp.sendRequest(reqBody, client.httpMethod)
+    req = HttpClientRequestRef.post(client.httpSession,
+                                    client.httpAddress.get,
+                                    body = reqBody.toOpenArrayByte(0, reqBody.len - 1))
+    res = await req.send()
 
-  if not res:
-    debug "Failed to send message to RPC server",
-          address = transp.remoteAddress(), msg_len = len(reqBody)
-    await transp.closeWait()
-    raise newException(ValueError, "Transport error")
-
-  debug "Message sent to RPC server", address = transp.remoteAddress(),
-        msg_len = len(reqBody)
+  debug "Message sent to RPC server",
+         address = client.httpAddress, msg_len = len(reqBody)
   trace "Message", msg = reqBody
+  echo "req body ", reqBody
 
-  let value = await transp.recvData(client.maxBodySize)
-  await transp.closeWait()
-  if value.len == 0:
-    raise newException(ValueError, "Empty response from server")
+  let resText = string.fromBytes(await res.getBodyBytes(client.maxBodySize))
+  trace "Response", text = resText
+  echo "response ", resText
```

Comment:
I think this `echo` should not be present here.

Reply:
does that also apply to [` echo "req body ", reqBody`](https://github.com/status-im/nim-json-rpc/blob/364dd2ebd132d7eb5e4516099250060b094c3daa/json_rpc/clients/httpclient.nim#L53)?

Reply:
https://github.com/bernardoaraujor/nim-json-rpc/commit/0b70c35a380d51d94a12e079e9f7316481e7a987